File: 5 Longest Palindromic Substring/PalindromicSubstring.java
import java.util.HashMap;
import java.util.Map;

class Solution {
   //bruteforce method
//    public String longestPalindrome(String s){
//     if(s.length() <=1 ){
//         return s;
//     }
//     int maxLen = 1;
//     String maxStr = s.substring(0, 1);
//     for (int i = 0; i<s.length(); i++){
//         for(int j= i+maxLen; j<s.length(); j++){
//             if(j-i > maxLen && isPalindrome(s.substring(i,j))){
//                 maxLen = j-i;
//                 maxStr=s.substring(i,j);
//             }
//         }
//     }
//     return maxStr;
//    }
//    private boolean isPalindrome(String str){
//     int left = 0;
//     int right = str.length() - 1;

//     while (left < right){
//         if(str.charAt(left) != str.charAt(right)){
//             return false;
//         }
//         left ++;
//         right --;

//     }
//     return true;


//Expand around center method
// public String longestPalindrome(String s){
//     if(s.length()<=1){
//         return s;
//     }
//     String maxStr = s.substring(0,1);
//     for(int i = 0; i<s.length()-1; i++){
//         String odd = expandFromCenter(s,i,i);
//         String even = expandFromCenter(s,i,i+1);

//         if(odd.length()>maxStr.length()){
//             maxStr = odd;
//         }
//         if (even.length()>maxStr.length()){
//             maxStr = even;
//         }
//     }
//     return maxStr;
// }


// private String expandFromCenter(String s, int left, int right){
//     while(left>=0 && right <s.length() && s.charAt(left) == s.charAt(right)){
//         left --;
//         right ++;
//     }
//     return s.substring(left+1, right);
// }
// ..................................................
// public String longestPalindrome(String s) {
//     int max =0;
//     if(s.length()==1){
//         return s;
//     }
//     String longestPalindromic = new String();
//     for(int i=0; i<s.length(); i++){
//         for(int j=i+1; j<s.length(); j++){
//             String aLongestPalindromic = s.substring(i, j);
//             if(isPalindromic(aLongestPalindromic) && max<aLongestPalindromic.length()){
//                 longestPalindromic = aLongestPalindromic;
//                 max = aLongestPalindromic.length();
//             }
//         }
//     }
//     return longestPalindromic;
// }
// public boolean isPalindromic(String sPassed){
//     int left =0, right=sPassed.length()-1;
//     while(left<right){
//         if(!(sPassed.charAt(left)==sPassed.charAt(right))){
//             return false;
//         }
//         left++;
//         right--;
//     }
//     return true;
// }

// public String longestPalindrome(String s) {
//         if(s.length()<=1){
//             return s;
//         }
//         String maxString = new String();
//         for(int i =0; i<s.length()-1; i++){
//             String odd = expandFromCenter(s, i, i);
//             String even = expandFromCenter(s,i,i+1);
//             if(odd.length()>maxString.length()){
//                 maxString = odd;
//             }
//             if(even.length()>maxString.length()){
//                 maxString=even;
//             }
//         }
//         return null;
// }
// private String expandFromCenter(String s, int left, int right){
//     while (left >=0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
//         left --;
//         right ++;
//     }
//     return s.substring(left+1, right);
// }
}

File: 5 Longest Palindromic Substring/PalindromicSubstring.py
# class Solution:
    # def longestPalindrome(self, s: str) -> str:
    #     if len(s) <=1:
    #         return s
        
    #     Max_Len=1
    #     Max_Str = s[0]
    #     for i in range(len(s)-1):
    #         for j in range(i+1, len(s)):
    #             if j-i+1>Max_Len and s[i:j+1] == s[i:j+1][::-1]:
    #                 Max_Len = j-i+1
    #                 Max_Str = s[i:j+1]
    #     return Max_Str
        # def longestPalindrome(self, s:str) -> str:
        #         if len(s) <=1:
        #                 return s
        #         def expand_from_center(left, right):
        #                 while left>=0 and right <len(s) and s[left] == s[right]:
        #                         left -=1
        #                         right +=1
        #                 return s[left+1:right]
        #         max_str = s[0]

        #         for i in range(len(s) -1):
        #                 odd = expand_from_center(i,i)
        #                 even = expand_from_center(i,i+1)

        #                 if len(odd) >len(max_str):
        #                         max_str = odd
        #                 if len(even) > len(max_str):
        #                         max_str = even
        #         return max_str

# class Solution:
#     def longestPalindrome(self, s:str) -> str:
#         if len(s) <=1:
#             return s
#         Max_Len = 1
#         Max_Str =s[0]
#         for i in range(len(s)-1):
#             for j in range(i+1, len(s)):
#                 if j-i+1>Max_Len and s[i:j+1] == s[i:j+1][::-1]:
#                     Max_Len = j-i+1
#                     Max_Str = s[i:j+1]

#         return Max_Str
    
class Solution:
    def longestPalindrome(self, s:str) -> str:
        def expand_from_center(left, right):
            while left>=0 and right<len(s) and s[left]==s[right]:
                right+=1
                left-=1
            return s[left+1:right]    

        if len(s)<=1:
            return s
        Max_Len = 0
        Max_Str = s[0]
        for i in range(len(s)-1):
            odd = expand_from_center(s,i,i)
            even = expand_from_center(s,i,i+1)
            if Max_Len<len(odd):
                    Max_Str = odd
                    Max_Len = len(odd)
            if Max_Len<len(even):
                 Max_Str = even
                 Max_Len = len(even)
        return Max_Str
                 



